### ISSUE TITLE
clangd can't find stdlib headers when run in darwin stdenv

### ISSUE THREAD

**Issue by maxbol** [NixOS/nixpkgs#354768]
### Describe the bug
When using the darwin stdenv, clangd does not seem to find headers to stdlib files. Using the /usr/bin/clangd that ships with xcode fixes the issue.

I would assume clangd should be able to find stdlib headers in the darwin stdenv. This seems to have been working prior to the big Apple SDK reimplementation in nixpkgs.

Building with gcc/clang/llvm etc works as expected. But LSP seems confused.

### Steps To Reproduce
0. Be a Mac user.
1. Clone repository https://github.com/maxbol/clangdarwin-minimal-reproduce
2. Enter a dev shell with `nix develop`
3. Run `clangd --check=./main.c`
4. Notice complaint about missing `stdlib.h`
5. Uncomment line in `flake.nix` that puts /usr/bin at the front of the path.
6. Repeat step 3.
7. Notice how everything works fine.

### Expected behavior
Clangd finds the stdlib headers in the darwin stdenv.

### Screenshots
With /nix/*/bin/clangd:
<img width="365" alt="image" src="https://github.com/user-attachments/assets/09028958-cd43-44a3-b9df-369dc821d556">

With /usr/bin/clangd:
<img width="365" alt="image" src="https://github.com/user-attachments/assets/f083b6fe-cbfc-46e2-a1e5-f0aaccefab77">

### Additional context
Worked as "expected" (could well be because of impurities that I was unknowingly depending on) until updating my local nix-profile to lastest nixos-unstable.

### Notify maintainers
@reckenrode - sorry if this is not about you, had a little difficulty understanding who was the maintainer of the darwin stdenv.

### Metadata

❯ nix-info -m
 - system: `"aarch64-darwin"`
 - host os: `Darwin 23.6.0, macOS 14.6.1`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.20.3`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by emilazy** [comment]
This may or may not be fixed on the current `staging-next` branch, if you’d like to test.

FWIW, @NixOS/darwin-core aka `lib.teams.darwin` maintains the Darwin `stdenv`, but you already pinged ⅓ of that team :)

**Comment by maxbol** [comment]
Thanks, I'll give it a try. And thanks for letting me know who to ping next time!

Also, I updated the steps to reproduce so you don't have to have any text editors installed (with all the impurities that entails...). :)

**Comment by maxbol** [comment]
It works!! 🏅

**Comment by reckenrode** [comment]
Thanks for the test case. It’s really helpful. 😄

I was also going to suggest checking staging-next, but I see that you have, and the SDK headers issue appears resolved. I assume that is due to the sysroot changes @emilazy made that are working their way through staging-next.

The issue with `stdbool.h` is that the `clangd` in the default environment is unwrapped. If I add `clang-tools` to the dev shell’s `packages`, it’s wrapped to interpret `NIX_CFLAGS_COMPILE`. Using that `clangd`, it can find `stdbool.h`.

**Comment by maxbol** [comment]
Yes, this works for me. I just simply didn't know I was supposed to do that, hehe. Huge thanks!

Sidenote: At some point I want to create some kind of repository with a bunch of boilerplate devshells for different environments and programming languages. I think it's one of the things that is unneccessarily hard to google. All these little quirks and extras necessary to get pretty standard functionality in place when using nix as a dev environment can be infuriating.

**Comment by emilazy** [comment]
FWIW I would like to merge `clang-tools`/`clang-analyzer` into just the normal `clang` package. So hopefully soon you won’t have to know that!

**Comment by skittishdev** [comment]
Hello,
I'm facing the same issue with `stdbool.h` not found for MacOS. I tried using `staging-next`, and tried including `clang-tools` as suggested in the comments above, but that doesn't seem to help. 

My `compile_commands.json` seems to indicate that clang is wrapped? But maybe I'm misunderstanding the comments above.

### compile_commands.json (excerpt)

```json
{
  "directory": "<path>/build/tests",
  "command": "/nix/store/zzx6cfl99zknfrj9v7lmrshwzslfjv26-clang-wrapper-19.1.7/bin/clang  -I<path>/include  -arch arm64 -isysroot /nix/store/rhlyld20zdzdi5wghb4b12gw52s9whr9-apple-sdk-11.3/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -mmacosx-version-min=11.3   -I/nix/store/4imrr3x1b5c3yprk191lpnrb57s706cv-criterion-2.4.2-dev/include -o CMakeFiles/test_alloc_and_free.dir/test_alloc_and_free.c.o -c <path>/tests/test_alloc_and_free.c",
  "file": "<path>/tests/test_alloc_and_free.c",
  "output": "tests/CMakeFiles/test_alloc_and_free.dir/test_alloc_and_free.c.o"
}
```

### My flake.nix

```nix
{
  inputs = {
    nixpkgs = {
      url = "github:NixOS/nixpkgs/staging-next";
    };
  };
  outputs = {
    self,
    nixpkgs,
    ...
  }@inputs: let
    inherit (self) outputs;

    forAllSystems = nixpkgs.lib.genAttrs [
      "x86_64-linux"
      "aarch64-darwin"
    ];
  in {
    devShells = forAllSystems (
      system: let
        pkgs = nixpkgs.legacyPackages.${system};
      in {
        default = pkgs.mkShell {
          nativeBuildInputs = [
            pkgs.clang_20
            pkgs.clang-tools
            pkgs.lldb
            pkgs.cmake
            pkgs.pkg-config
            pkgs.criterion.dev
          ];
        };
      }
    );
  };
}

```

**Comment by maxbol** [comment]
>     devShells = forAllSystems (
>       system: let
>         pkgs = nixpkgs.legacyPackages.${system};
>       in {
>         default = pkgs.mkShell {
>           nativeBuildInputs = [
>             pkgs.clang_20
>             pkgs.clang-tools
>             pkgs.lldb
>             pkgs.cmake
>             pkgs.pkg-config
>             pkgs.criterion.dev
>           ];
>         };
>       }
>     );
>   };
> }
> 
> ```

Not sure if related but I put clang-tools in packages, not nativeBuildInputs.
